README.md
[![Build Status](https://travis-ci.org/glesys/docker-machine-driver-glesys.svg?branch=master)](https://travis-ci.org/glesys/docker-machine-driver-glesys)

# GleSYS driver for Docker Machine

This is the official Docker Machine driver to create Docker machines with
[GleSYS](https://glesys.com).

## Installation

### From a Release

Binaries are available for Linux, OS X and Windows on the
[releases](https://github.com/glesys/docker-machine-driver-glesys/releases)
page.

Download the binary for your platform, make sure to add it to your `PATH` and
make it executable.

Rename the file to `docker-machine-driver-glesys`.

#### Windows

Rename the file to `docker-machine-driver-glesys.exe`

### Building from Source

Make sure you have Go and `GOPATH` setup correctly. Then run:

```shell
go get github.com/glesys/docker-machine-driver-glesys
cd $GOPATH/src/github.com/glesys/docker-machine-driver-glesys
go install ./cmd/docker-machine-driver-glesys
```

## Obtaining Credentials

To use this driver you need a GleSYS Cloud account and a valid API key. You can
sign up for a free account at https://glesys.com/signup then visit
https://customer.glesys.com to create an API key for any of your projects.

## Using the Driver

You can ensure that `docker-machine` can find the GleSYS driver by asking for
the driver help:

```shell
docker-machine create -d glesys | grep glesys

  --glesys-project                             GleSYS project (e.g. CL12345) [$GLESYS_PROJECT]
  --glesys-api-key                             GleSYS API key [$GLESYS_API_KEY]

  --glesys-bandwidth "100"                     Bandwidth in MBit/s
  --glesys-campaign-code                       Campaign code to use for the machine
  --glesys-cpu "2"                             Number of CPU cores
  --glesys-data-center "Falkenberg"            Data center to place the machine in
  --glesys-memory "2048"                       Memory in MB
  --glesys-root-password                       Root password to use for the machine. If omitted, a random password will be generated (VMware only)
  --glesys-username-kvm "docker-machine"       Username to use in KVM platform
  --glesys-ssh-key-path                        Path to the SSH key file you want to use. If omitted, a new key will be generated
  --glesys-storage "20"                        Storage in GB
  --glesys-template "Ubuntu 16.04 LTS 64-bit"  Template to use for the machine
  --glesys-platform "VMware"                   Virtualization platform (VMware or KVM)
```

To create a machine you need to specify a project and an API key:

```shell
docker-machine create -d glesys --glesys-project=CL12345 --glesys-api-key=my-api-key example-host

Running pre-create checks...
Creating machine...
(example-host) Waiting for machine to come online...
Waiting for machine to be running, this may take a few minutes...
Detecting operating system of created instance...
Waiting for SSH to be available...
Detecting the provisioner...
Provisioning with ubuntu(systemd)...
Installing Docker...
Copying certs to the local machine directory...
Copying certs to the remote machine...
Setting Docker configuration on the remote daemon...
Checking connection to Docker...
Docker is up and running!
To see how to connect your Docker Client to the Docker Engine running on this virtual machine,
run: docker-machine env example-host
```

## Help

If you need any help using this driver feel free to send an email to
support@glesys.com.

## License

The contents of this repository are distributed under the MIT license, see
[LICENSE](LICENSE).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [1.2.0] - 2023-02-27
- Update default OS template
- Update Windows specifics in README
- Update glesys-go dependency
- Bump dependencies
## [1.1.0] - 2020-05-11
### Added
- Support for KVM platform

## [1.0.0] - 2017-02-23
### Added
- Initial release


glesys_test.go
package glesys

import (
	"github.com/docker/machine/libmachine/drivers"
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestConfigFromFlags(t *testing.T) {
	driver := NewDriver("default", "path")

	checkFlags := &drivers.CheckDriverOptions{
		FlagsValues: map[string]interface{}{
			"glesys-api-key":       "TOKEN",
			"glesys-bandwidth":     100,
			"glesys-campaign-code": "CODE",
			"glesys-cpu":           2,
			"glesys-data-center":   "Falkenberg",
			"glesys-storage":       40,
			"glesys-memory":        2048,
			"glesys-project":       "cl12345",
			"glesys-root-password": "secret-password",
			"glesys-template":      "Debian 8 64-bit",
			"glesys-ssh-key-path":  "~/.ssh/id_rsa",
		},
		CreateFlags: driver.GetCreateFlags(),
	}

	err := driver.SetConfigFromFlags(checkFlags)

	assert.NoError(t, err)
	assert.Empty(t, checkFlags.InvalidFlags)
}

func TestConfigRequiresProject(t *testing.T) {
	driver := NewDriver("default", "path")

	checkFlags := &drivers.CheckDriverOptions{
		FlagsValues: map[string]interface{}{
			"glesys-api-key": "TOKEN",
		},
		CreateFlags: driver.GetCreateFlags(),
	}

	err := driver.SetConfigFromFlags(checkFlags)
	assert.EqualError(t, err, "glesys driver requires the --glesys-project option")
}

func TestConfigRequiresApiKey(t *testing.T) {
	driver := NewDriver("default", "path")

	checkFlags := &drivers.CheckDriverOptions{
		FlagsValues: map[string]interface{}{
			"glesys-project": "cl12345",
		},
		CreateFlags: driver.GetCreateFlags(),
	}

	err := driver.SetConfigFromFlags(checkFlags)
	assert.EqualError(t, err, "glesys driver require the --glesys-api-key option")
}

func TestStringToEnvVar(t *testing.T) {
	assert.Equal(t, "GLESYS_API_KEY", stringFlagToEnvVar("glesys-api-key"), "return value is correct")
	assert.Equal(t, "GLESYS_PROJECT", stringFlagToEnvVar("glesys-project"), "return value is correct")
}

func TestDriverName(t *testing.T) {
	driver := NewDriver("default", "path")

	assert.Equal(t, "glesys", driver.DriverName(), "Driver name is correct")
}

func TestSSHPort(t *testing.T) {
	driver := NewDriver("default", "path")
	port, err := driver.GetSSHPort()

	assert.NoError(t, err, "there is no error")
	assert.Equal(t, 22, port, "SSH port is correct")
}

func TestSSHUsername(t *testing.T) {
	driver := NewDriver("default", "path")

	assert.Equal(t, "root", driver.GetSSHUsername(), "SSH username is correct")
}

func TestGeneratePassword(t *testing.T) {
	password := generatePassword(64)

	assert.Len(t, password, 64, "password has correct length")
}

func TestDriver_PreCreateCheck(t *testing.T) {
	driver := NewDriver("default", "path")
	err := driver.PreCreateCheck()
	assert.NoError(t, err)
}


glesys.go
package glesys

import (
	"context"
	"fmt"
	"github.com/glesys/glesys-go/v6"
	"io/ioutil"
	"math/rand"
	"net"
	"os"
	"strings"
	"time"

	"github.com/docker/machine/libmachine/drivers"
	"github.com/docker/machine/libmachine/log"
	"github.com/docker/machine/libmachine/mcnflag"
	"github.com/docker/machine/libmachine/ssh"
	"github.com/docker/machine/libmachine/state"
)

const (
	defaultBandwidth   = 100
	defaultCPU         = 2
	defaultDataCenter  = "Falkenberg"
	defaultMemory      = 2048
	defaultStorage     = 20
	defaultTemplate    = "Ubuntu 20.04 LTS 64-bit"
	defaultPlatform    = "VMware"
	defaultUsernameKvm = "docker-machine"
)

const (
	apiKeyFlag       = "glesys-api-key"
	bandwidthFlag    = "glesys-bandwidth"
	campaignCodeFlag = "glesys-campaign-code"
	cpuFlag          = "glesys-cpu"
	dataCenterFlag   = "glesys-data-center"
	memoryFlag       = "glesys-memory"
	projectFlag      = "glesys-project"
	rootPasswordFlag = "glesys-root-password"
	sshKeyPathFlag   = "glesys-ssh-key-path"
	storageFlag      = "glesys-storage"
	templateFlag     = "glesys-template"
	platformFlag     = "glesys-platform"
	usernameKVMFlag  = "glesys-username-kvm"
)

// Driver for GleSYS
type Driver struct {
	*drivers.BaseDriver
	APIKey       string
	Bandwidth    int
	CampaignCode string
	CPU          int
	DataCenter   string
	Memory       int
	Project      string
	RootPassword string
	ServerID     string
	Storage      int
	Template     string
	Platform     string
	UsernameKVM  string
}

// NewDriver creates a new driver
func NewDriver(hostName, storePath string) drivers.Driver {
	return &Driver{
		BaseDriver: &drivers.BaseDriver{
			MachineName: hostName,
			StorePath:   storePath,
		},
	}
}

func stringFlagToEnvVar(s string) string {
	return strings.ToUpper(strings.Replace(s, "-", "_", -1))
}

// GetCreateFlags defines all flags and environment variables that can be used with docker-machine create
func (d *Driver) GetCreateFlags() []mcnflag.Flag {
	return []mcnflag.Flag{
		mcnflag.StringFlag{
			EnvVar: stringFlagToEnvVar(projectFlag),
			Name:   projectFlag,
			Usage:  "GleSYS project (e.g. CL12345)",
		},
		mcnflag.StringFlag{
			EnvVar: stringFlagToEnvVar(apiKeyFlag),
			Name:   apiKeyFlag,
			Usage:  "GleSYS API key",
		},
		mcnflag.IntFlag{
			Name:  memoryFlag,
			Usage: "Memory in MB",
			Value: defaultMemory,
		},
		mcnflag.IntFlag{
			Name:  cpuFlag,
			Usage: "Number of CPU cores",
			Value: defaultCPU,
		},
		mcnflag.IntFlag{
			Name:  storageFlag,
			Usage: "Storage in GB",
			Value: defaultStorage,
		},
		mcnflag.IntFlag{
			Name:  bandwidthFlag,
			Usage: "Bandwidth in MBit/s",
			Value: defaultBandwidth,
		},
		mcnflag.StringFlag{
			Name:  dataCenterFlag,
			Usage: "Data center to place the machine in",
			Value: defaultDataCenter,
		},
		mcnflag.StringFlag{
			Name:  rootPasswordFlag,
			Usage: "Root password to use for the machine. If omitted, a random password will be generated (VMware only)",
		},
		mcnflag.StringFlag{
			Name:  templateFlag,
			Usage: "Template to use for the machine",
			Value: defaultTemplate,
		},
		mcnflag.StringFlag{
			Name:  campaignCodeFlag,
			Usage: "Campaign code to use for the machine",
			Value: "",
		},
		mcnflag.StringFlag{
			Name:  sshKeyPathFlag,
			Usage: "Path to the SSH key file you want to use. If omitted, a new key will be generated",
			Value: "",
		},
		mcnflag.StringFlag{
			Name:  platformFlag,
			Usage: "Virtualization platform (VMware or KVM)",
			Value: defaultPlatform,
		},
		mcnflag.StringFlag{
			Name:  usernameKVMFlag,
			Usage: "Username to use in KVM platform",
			Value: defaultUsernameKvm,
		},
	}
}

// SetConfigFromFlags configures the driver with the object that was returned
// by RegisterCreateFlags
func (d *Driver) SetConfigFromFlags(opts drivers.DriverOptions) error {
	d.Project = opts.String(projectFlag)
	d.APIKey = opts.String(apiKeyFlag)

	if d.Project == "" {
		return fmt.Errorf("glesys driver requires the --%v option", projectFlag)
	}

	if d.APIKey == "" {
		return fmt.Errorf("glesys driver require the --%v option", apiKeyFlag)
	}

	d.Bandwidth = opts.Int(bandwidthFlag)
	d.CampaignCode = opts.String(campaignCodeFlag)
	d.CPU = opts.Int(cpuFlag)
	d.DataCenter = opts.String(dataCenterFlag)
	d.Memory = opts.Int(memoryFlag)
	d.RootPassword = opts.String(rootPasswordFlag)
	d.SSHKeyPath = opts.String(sshKeyPathFlag)
	d.Storage = opts.Int(storageFlag)
	d.Template = opts.String(templateFlag)
	d.Platform = opts.String(platformFlag)
	d.UsernameKVM = opts.String(usernameKVMFlag)

	if d.Platform != "VMware" && d.Platform != "KVM" {
		return fmt.Errorf("platform %v is not valid, supported platforms are VMware and KVM", d.Platform)
	}

	return nil
}

// PreCreateCheck allows for pre-create operations to make sure a driver is ready for creation
func (d *Driver) PreCreateCheck() error {
	if d.RootPassword == "" {
		d.RootPassword = generatePassword(64)
	}

	if d.SSHKeyPath != "" {
		if _, err := os.Stat(d.SSHKeyPath); os.IsNotExist(err) {
			return fmt.Errorf("SSH key file does not exist: %q", d.SSHKeyPath)
		}
	}

	return nil
}

// Create a host using the driver's config
func (d *Driver) Create() error {
	client := d.getClient()

	if d.SSHKeyPath == "" {
		if err := ssh.GenerateSSHKey(d.GetSSHKeyPath()); err != nil {
			return err
		}
	}

	publicKey, err := ioutil.ReadFile(d.GetSSHKeyPath() + ".pub")
	if err != nil {
		return err
	}

	serverParams := glesys.CreateServerParams{
		Bandwidth:    d.Bandwidth,
		CampaignCode: d.CampaignCode,
		CPU:          d.CPU,
		DataCenter:   d.DataCenter,
		Hostname:     d.GetMachineName(),
		IPv4:         "any",
		IPv6:         "any",
		Memory:       d.Memory,
		Platform:     d.Platform,
		Storage:      d.Storage,
		Template:     d.Template,
	}
	if d.Platform == "KVM" {
		publicKeys := []string{
			string(publicKey),
		}
		serverParams = serverParams.WithUser(d.UsernameKVM, publicKeys, "")
	} else {
		serverParams.Password = d.RootPassword
		serverParams.PublicKey = string(publicKey)
	}

	server, err := client.Servers.Create(context.Background(), serverParams)

	if err != nil {
		return fmt.Errorf("Failed to create machine: %v", err)
	}

	d.ServerID = server.ID
	d.IPAddress = server.IPList[0].Address

	return nil
}

// DriverName returns the name of the driver
func (d *Driver) DriverName() string {
	return "glesys"
}

// GetSSHHostname returns hostname for use with ssh
func (d *Driver) GetSSHHostname() (string, error) {
	return d.GetIP()
}

// GetSSHPort returns port for use with ssh
func (d *Driver) GetSSHPort() (int, error) {
	return 22, nil
}

// GetSSHUsername returns username for use with ssh
func (d *Driver) GetSSHUsername() string {
	if d.Platform == "KVM" {
		return d.UsernameKVM
	} else {
		return "root"
	}

}

// GetURL returns a Docker compatible URL for connecting to this machine
// e.g. tcp://1.2.3.4:2376
func (d *Driver) GetURL() (string, error) {
	if err := drivers.MustBeRunning(d); err != nil {
		return "", err
	}

	ip, err := d.GetIP()
	if err != nil {
		return "", err
	}

	return fmt.Sprintf("tcp://%s", net.JoinHostPort(ip, "2376")), nil
}

// GetState returns the current state of a machine
func (d *Driver) GetState() (state.State, error) {

	s, err := d.getClient().Servers.Details(context.Background(), d.ServerID)
	if err != nil {
		return state.Error, err
	}
	switch s.State {
	case "locked":
		return state.None, nil
	case "running":
		return state.Running, nil
	case "stopped":
		return state.Stopped, nil
	}

	return state.None, nil
}

// Kill stops a machine forcefully
func (d *Driver) Kill() error {
	return d.getClient().Servers.Stop(context.Background(), d.ServerID, glesys.StopServerParams{
		Type: "hard",
	})
}

func generatePassword(passlen int) string {
	rand.Seed(time.Now().UTC().UnixNano())
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	result := make([]byte, passlen)
	for i := 0; i < passlen; i++ {
		result[i] = chars[rand.Intn(len(chars))]
	}

	return string(result)
}

// Remove a machine
func (d *Driver) Remove() error {
	if d.ServerID == "" {
		log.Warn("Can not remove the machine, ID is missing")
		return nil
	}
	return d.getClient().Servers.Destroy(context.Background(), d.ServerID, glesys.DestroyServerParams{
		KeepIP: false,
	})
}

// Restart a machine.
func (d *Driver) Restart() error {
	return d.getClient().Servers.Stop(context.Background(), d.ServerID, glesys.StopServerParams{
		Type: "reboot",
	})
}

// Start a machine
func (d *Driver) Start() error {
	return d.getClient().Servers.Start(context.Background(), d.ServerID)
}

// Stop a machine gracefully
func (d *Driver) Stop() error {
	return d.getClient().Servers.Stop(context.Background(), d.ServerID, glesys.StopServerParams{
		Type: "soft",
	})
}

func (d *Driver) getClient() *glesys.Client {
	return glesys.NewClient(d.Project, d.APIKey, "docker-machine-driver-glesys/1.1.0")
}
